File: scripts/http-probe.mjs
// SPDX-License-Identifier: Apache-2.0
import {readFileSync,existsSync} from 'node:fs';
import {HttpMcpClient} from '@psp-cdl/mcpproxy/mcp';
import {McpDispatchGate} from '@psp-cdl/mcpproxy';
import {fixture} from './dispatch-fixtures.mjs';
import {suite,runHttpCase,approval} from './http-fixtures.mjs';
if(process.argv[2]==='--report') {
  const result=[];for(const c of suite.cases)result.push(await runHttpCase(c));console.log(JSON.stringify(result));
}else {
  const c=JSON.parse(process.argv[2]);let peer,f;
  try {
    peer=await HttpMcpClient.connect({endpoint:c.endpoint,allowLoopbackHttp:!c.rejectHttp,serverInfo:{name:'psp-cdl-reference',version:'0.1.0'},timeoutMs:700,...(c.ca?{caPem:readFileSync(c.ca,'utf8')}:{})},resource=>{if(resource!==c.endpoint)throw Error('resource mismatch');return c.token??'test-downstream';});
    f=await fixture(c.settings??{});
    const gate=new McpDispatchGate(f.store,f.host,'registry-1',peer.registrations('echo',[approval],()=>1000));
    const options={deadline:1800,cancelled:()=>!!c.cancel&&existsSync(c.spy)&&readFileSync(c.spy,'utf8').includes('call')};
    const result=await gate.callTool('test-owner',f.session.sessionId,{name:'echo.read',arguments:{message:'hello 🧪'}},options);
    console.log(JSON.stringify({code:'OK',released:1,data:result.data,trustLevel:result.provenance.trustLevel}));
  }catch(e){console.log(JSON.stringify({code:e.code??'UNEXPECTED_ERROR',released:0}));}
  finally{await peer?.close();f?.close();}
}

File: scripts/http_probe.py
# SPDX-License-Identifier: Apache-2.0
import json
import sys
from pathlib import Path
from http_fixtures import SUITE, run_http_case, APPROVAL
from dispatch_fixtures import Fixture
from psp_cdl_mcpproxy import McpDispatchGate
from psp_cdl_mcpproxy.mcp import HttpMcpClient

def probe(c):
    peer=f=None
    try:
        def credential(resource):
            assert resource==c["endpoint"]
            return c.get("token","test-downstream")
        peer=HttpMcpClient.connect({"endpoint":c["endpoint"],"allowLoopbackHttp":not c.get("rejectHttp",False),"serverInfo":{"name":"psp-cdl-reference","version":"0.1.0"},"timeoutMs":700,**({"caPem":Path(c["ca"]).read_text()} if c.get("ca") else {})},credential)
        f=Fixture(c.get("settings"))
        gate=McpDispatchGate(f.store,f,"registry-1",peer.registrations("echo",[APPROVAL],f.now))
        cancelled=lambda:bool(c.get("cancel") and Path(c["spy"]).exists() and "call" in Path(c["spy"]).read_text())
        result=gate.call_tool("test-owner",f.session["sessionId"],{"name":"echo.read","arguments":{"message":"hello 🧪"}},{"deadline":1800,"cancelled":cancelled})
        return {"code":"OK","released":1,"data":result["data"],"trustLevel":result["provenance"]["trustLevel"]}
    except Exception as exc: return {"code":getattr(exc,"code","UNEXPECTED_ERROR"),"released":0}
    finally:
        if peer: peer.close()
        if f: f.close()

if __name__=="__main__":
    print(json.dumps([run_http_case(c) for c in SUITE["cases"]] if sys.argv[1]=="--report" else probe(json.loads(sys.argv[1])),ensure_ascii=False))
